lib/Feature/FeatureModelTransaction.cpp

```diff
@@ -36,6 +40,86 @@ void removeFeature(FeatureModel &FM,
   Trans.commit();
 }
 
+bool fvIsLeave(FeatureModel &FM, detail::FeatureVariantTy &FV) {
+  Feature *ActualFeature;
+  std::visit(
+      Overloaded{[&ActualFeature, &FM](Feature *F) { ActualFeature = F; },
+                 [&ActualFeature, &FM](string &F) {
+                   ActualFeature = FM.getFeature(F);
+                 }},
+      FV);
+  return ActualFeature->isLeave();
+}
+
+// TODO: delete, if other approach is used
+/*bool hasSuccessorsAmongOthers(
+    FeatureModel &FM, detail::FeatureVariantTy &FV,
+    std::vector<detail::FeatureVariantTy>::iterator Begin,
+    std::vector<detail::FeatureVariantTy>::iterator End) {
+  Feature *ActualFeature;
+  std::visit(
+      Overloaded{[&ActualFeature, &FM](Feature *F) { ActualFeature = F; },
+                 [&ActualFeature, &FM](string &F) {
+                   ActualFeature = FM.getFeature(F);
+                 }},
+      FV);
+
+  if (ActualFeature->children().empty()) {
+    return false;
+  }
+
+  for (auto Child : ActualFeature->children()) {
+    if (std::find(Begin, End, Child) != End) {
+    }
+  }
+  return false;
+}*/
+
+void removeFeatures(FeatureModel &FM,
+                    std::vector<detail::FeatureVariantTy> FeaturesToBeDeleted,
+                    bool Recursive) {
+  removeFeatures(FM, FeaturesToBeDeleted.begin(), FeaturesToBeDeleted.end(),
+                 Recursive);
+}
+
+void removeFeatures(FeatureModel &FM,
+                    std::vector<detail::FeatureVariantTy>::iterator Begin,
+                    std::vector<detail::FeatureVariantTy>::iterator End,
+                    bool Recursive) {
+  // if everything was deleted
+  if (Begin == End) {
+    return;
+  }
+
+  // partition FeaturesToBeDeleted into others and leaves --> remove leaves and
+  // recursively call this function on others
+  auto NoDeleteIt =
+      std::partition(Begin, End, [&FM](detail::FeatureVariantTy &FV) {
+        return !fvIsLeave(FM, FV); // TODO: enhance to also delete non-leaves in
+                                   // recursive=true mode
+      });
+
+  // check if leaves are present --> if not and return non-deletable features
+  if (NoDeleteIt == End) {
+    // TODO: return list of non-deletable Features?
+    return;
+  }
+
+  // remove leaves
+  auto Trans = FeatureModelModifyTransaction::openTransaction(FM);
+  auto it = NoDeleteIt;
+  while (it != End) {
+    Trans.removeFeature(*it, true);
```

Comment:
Should we not use `Recursive` here? When we pass `true` we always delete them recursively. On the other hand, we only remove leaves, so I guess, it does not matter.

Reply:
Done